pkg/clustermanager/kube_proxy.go

```diff
@@ -0,0 +1,505 @@
+		return up.CleanupAfterUpgrade(ctx, u.log, managementClusterClient, workloadClusterClient, spec)
+	})
+}
+
+func (u KubeProxyCLIUpgrader) buildClients(
+	managementClusterKubeconfigPath, workloadClusterKubeconfigPath string,
+) (managementClusterClient, workloadClusterClient client.Client, err error) {
+	u.log.V(4).Info("Building client for management cluster", "kubeconfig", managementClusterKubeconfigPath)
+	managementClusterClient, err = u.clientFactory.BuildClientFromKubeconfig(managementClusterKubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u.log.V(4).Info("Building client for workload cluster", "kubeconfig", workloadClusterKubeconfigPath)
+	workloadClusterClient, err = u.clientFactory.BuildClientFromKubeconfig(workloadClusterKubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return managementClusterClient, workloadClusterClient, nil
+}
+
+// NewKubeProxyUpgrader builds a new KubeProxyUpgrader.
+func NewKubeProxyUpgrader(opts ...KubeProxyUpgraderOpt) KubeProxyUpgrader {
+	u := &KubeProxyUpgrader{
+		updateKubeProxyRetries: 30,
+		updateKubeProxyBackoff: 2 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(u)
+	}
+
+	return *u
+}
+
+// KubeProxyUpgrader prepares a cluster for a kube-proxy upgrade.
+type KubeProxyUpgrader struct {
+	updateKubeProxyRetries int
+	updateKubeProxyBackoff time.Duration
+}
+
+// KubeProxyUpgraderOpt allows to customize a KubeProxyUpgraderOpt
+// on construction.
+type KubeProxyUpgraderOpt func(*KubeProxyUpgrader)
+
+// WithUpdateKubeProxyTiming allows to customize the retry paramenter for the
+// kube-proxy version update. This is for unit tests.
+func WithUpdateKubeProxyTiming(retries int, backoff time.Duration) KubeProxyUpgraderOpt {
+	return func(u *KubeProxyUpgrader) {
+		u.updateKubeProxyRetries = retries
+		u.updateKubeProxyBackoff = backoff
+	}
+}
+
+// PrepareForUpgrade gets the workload cluster ready for a smooth transition between the
+// old kube-proxy that always uses iptables legacy and the new one that detects the host preference
+// and is able to work with nft as well. This is idempotent, so it can be called in a loop if transient
+// errors are a risk.
+func (u KubeProxyUpgrader) PrepareForUpgrade(ctx context.Context, log logr.Logger, managementClusterClient, workloadClusterClient client.Client, spec *cluster.Spec) error {
+	kcp, err := controller.KubeadmControlPlane(ctx, managementClusterClient, spec.Cluster)
+	if err != nil {
+		return errors.Wrap(err, "reading the kubeadm control plane for an upgrade")
+	}
+
+	_, newVersion := oci.Split(spec.VersionsBundle.KubeDistro.KubeProxy.URI)
+
+	// If the new spec doesn't include the new kube-proxy or if the current cluster already has it, skip this
+	if needsPrepare, err := needsKubeProxyPreUpgrade(spec, kcp); err != nil {
+		return err
+	} else if !needsPrepare {
+		log.V(4).Info("Kube-proxy upgrade doesn't need special handling", "currentVersion", kcp.Spec.Version, "newVersion", newVersion)
```

Comment:
Isn't this just the following? Or did you want to highlight the special handling between old to new differently?
```suggestion
		log.V(4).Info("skipping kube-proxy upgrade setup as it contains same version", "currentVersion", kcp.Spec.Version, "newVersion", newVersion)
```

Reply:
It doesn't necessarily contain the same version
It can contain a different version but none of them including the new kube-proxy
Or it can also contain two different versions but both including the new kube-proxy

Reply:
To clarify: this prepare process is only necessary when there is a transition between a kube-proxy that always uses legacy (the "old" one) and a kube-proxy that uses the new iptables wrapper (the "new" one), which adds support for nft.